hbase-server/src/main/java/org/apache/hadoop/hbase/replication/HBaseReplicationEndpoint.java

```diff
@@ -87,8 +108,24 @@ public void stop() {
   }
 
   @Override
-  protected void doStart() {
+  protected synchronized void doStart() {
     try {
+      try (Admin admin = getPeerConnection().getAdmin()) {
+        String version = admin.getClusterMetrics().getHBaseVersion();
+        LOG.info("Peer cluster version is {} for peer {}", version, ctx.getPeerId());
+        if (Integer.parseInt(version.split("\\.")[0]) >= 3) {
```

Comment:
Thanks @wchevreuil for reviewing.
[ReplicationSinkManager.chooseSinks](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSinkManager.java#L151) will be called to re-fetch the sinks when HBaseInterClusterReplicationEndpoint.replicate [catches ConnectException or UnknownHostException](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/HBaseInterClusterReplicationEndpoint.java#L563), It's same as before.
And I've added a config "hbase.replication.fetch.servers.usezk" to optionally still use ZK impl.

Reply:
> ReplicationSinkManager.chooseSinks will be called to re-fetch the sinks when HBaseInterClusterReplicationEndpoint.replicate catches ConnectException or UnknownHostException, It's same as before.

Yeah, haven't looked at the outer catch from replicate. Another thought them is if we really need the new chore? I wonder how effective it would really be to avoid trying stale sinks.

Reply:
> Another thought them is if we really need the new chore? I wonder how effective it would really be to avoid trying stale sinks.

Thank @wchevreuil  for thinking about this. 
I had a thought, there is some differences between [regionServers](https://github.com/apache/hbase/blob/0b85729da406df5ca33fd5c11d8cfe5b2bb16a41/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/HBaseReplicationEndpoint.java#L53) stored in HBaseReplicationEndpoint and [sinks](https://github.com/apache/hbase/blob/0b85729da406df5ca33fd5c11d8cfe5b2bb16a41/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSinkManager.java#L80) stored in ReplicationSinkManager. The latter is a random part of the former. The new chore can remove abandoned servers and discover new servers. If there is no such chore, new servers can't be sinks unless exceptions happen due to RSs crash.